paddle/fluid/framework/details/multi_devices_graph_builder.cc

```diff
@@ -207,7 +131,97 @@ std::unique_ptr<SSAGraph> MultiDevSSAGraphBuilder::Build(
   }
 
   return std::unique_ptr<SSAGraph>(graph);
-}  // namespace details
+}
+
+void MultiDevSSAGraphBuilder::InsertNCCLAllReduceOp(
+    SSAGraph *result, const std::string &og) const {
+#ifdef PADDLE_WITH_CUDA
+  result->ops_.emplace_back(
+      new NCCLAllReduceOpHandle(local_scopes_, places_, *nccl_ctxs_));
+  auto *op_handle = result->ops_.back().get();
+
+  for (size_t i = 0; i < places_.size(); ++i) {
+    auto &p = places_[i];
+    auto &vars = result->vars_[i][og];
+    if (vars.empty()) {  // This device has no data. continue.
+      continue;
+    }
```

Comment:
I am curious about why `vars` could be empty, `nccl_allReduce_op_handle` will do `allReduce` on the tensors of `local_scope` according to the name of inputs_[0].

https://github.com/PaddlePaddle/Paddle/blob/9ced6da444ac759f755b9d7521a9d4882afabd1d/paddle/fluid/framework/details/nccl_all_reduce_op_handle.cc#L68-L80

Could you give an example?